packages/uploadthing/src/client.ts

```diff
@@ -312,3 +260,137 @@ export const generateClientDropzoneAccept = (fileTypes: string[]) => {
 };
 
 export { resolveMaybeUrlArg };
+
+async function uploadMultipart(
+  file: File,
+  presigned: MPUResponse,
+  opts: {
+    reportEventToUT: ReturnType<typeof createUTReporter>;
+    onUploadProgress?: UploadFilesOptions<any, any>["onUploadProgress"];
+  },
+) {
+  let etags: { tag: string; partNumber: number }[];
+  let uploadedBytes = 0;
+
+  try {
+    etags = await Promise.all(
+      presigned.presignedUrls.map(async (url, index) => {
+        const offset = presigned.chunkSize * index;
+        const end = Math.min(offset + presigned.chunkSize, file.size);
+        const chunk = file.slice(offset, end);
+
+        const etag = await uploadPartWithProgress({
+          url,
+          chunk: chunk,
+          contentDisposition: presigned.contentDisposition,
+          fileType: file.type,
+          fileName: file.name,
+          maxRetries: 10,
+          onProgress: (delta) => {
+            uploadedBytes += delta;
+            const percent = (uploadedBytes / file.size) * 100;
+            opts.onUploadProgress?.({ file: file.name, progress: percent });
+          },
+        });
+
+        return { tag: etag, partNumber: index + 1 };
+      }),
+    );
+  } catch (error) {
+    await opts.reportEventToUT("failure", {
+      fileKey: presigned.key,
+      uploadId: presigned.uploadId,
+      fileName: file.name,
+      s3Error: (error as Error).toString(),
+    });
+    throw "unreachable"; // failure event will throw for us
+  }
+
+  // Tell the server that the upload is complete
+  const uploadOk = await opts.reportEventToUT("multipart-complete", {
+    uploadId: presigned.uploadId,
+    fileKey: presigned.key,
+    etags,
+  });
+  if (!uploadOk) {
+    console.log("Failed to alert UT of upload completion");
+    throw new UploadThingError({
+      code: "UPLOAD_FAILED",
+      message: "Failed to alert UT of upload completion",
+    });
+  }
+}
+
+async function uploadPresignedPost(
+  file: File,
+  presigned: PSPResponse,
+  opts: {
+    onUploadProgress?: UploadFilesOptions<any, any>["onUploadProgress"];
+  },
+) {
+  const formData = new FormData();
+  Object.entries(presigned.fields).forEach(([k, v]) => formData.append(k, v));
+
+  // Give content type to blobs because S3 is dumb
+  // check if content-type is one of the allowed types, or if not and blobs are allowed, use application/octet-stream
+  if (
+    presigned.fileType === file.type.split("/")[0] ||
+    presigned.fileType === file.type
+  ) {
+    formData.append("Content-Type", file.type);
+  } else if (presigned.fileType === "blob") {
+    formData.append("Content-Type", "application/octet-stream");
+  } else if (presigned.fileType === "pdf") {
+    formData.append("Content-Type", "application/pdf");
+  }
```

Comment:
should this logic go on infra side and appended on the fields there?

Reply:
Yeah, not sure why we hadn't done it that way in the past